pdr_utils/contract.py

```diff
@@ -285,21 +327,70 @@ def soonest_block_to_predict(self, block):
         return self.contract_instance.functions.soonestBlockToPredict(block).call()
 
     def submit_prediction(
-        self, predicted_value, stake_amount, prediction_block, wait_for_receipt=True
+        self,
+        predicted_value: bool,
+        stake_amount: int,
+        prediction_ts: int,
+        wait_for_receipt=True,
     ):
-        """Sumbits a prediction"""
-        # TO DO - check allowence first, only do approve if needed
+        """
+        Submits a prediction with the specified stake amount to the smart contract on the blockchain.
+
+        @param predicted_value: The predicted value (True or False)
+        @param stake_amount: The amount of ETH to be staked on the prediction
+        @param prediction_ts: The prediction timestamp, must be the start timestamp of a candle.
+        @param wait_for_receipt: If True, the function waits for the transaction receipt after submission.
+                                If False, it returns immediately after sending the transaction. Default is True.
+
+        @return: If wait_for_receipt is True, returns the transaction receipt.
+                If wait_for_receipt is False, returns the transaction hash immediately after sending the transaction.
+                If an exception occurs during the  process, returns None.
+        """
         amount_wei = self.config.w3.to_wei(str(stake_amount), "ether")
+
+        # Check allowance first, only approve if needed
+        current_allowance = self.token.allowance(
+            self.config.owner, self.contract_address
+        )
+        if current_allowance < amount_wei:
+            try:
+                self.token.approve(self.contract_address, amount_wei)
+            except Exception as e:
+                print("Error while approving the contract to spend tokens:", e)
+                return None
+
         self.token.approve(self.contract_address, amount_wei)
         gasPrice = self.config.w3.eth.gas_price
         try:
-            tx = self.contract_instance.functions.submitPredval(
-                predicted_value, amount_wei, prediction_block
-            ).transact({"from": self.config.owner, "gasPrice": gasPrice})
-            print(f"Submitted prediction, txhash: {tx.hex()}")
+            txhash = None
+            if is_sapphire_rpc(self.config.rpc_url):
```

Comment:
we should check based on chainId.
Maybe somebody is running their own node and then url is not a match